drivers/mcp23x17/mcp23x17.c

```diff
@@ -0,0 +1,578 @@
+        if (gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_FALLING,
+                                               _irq_isr, (void*)dev)) {
+            DEBUG_DEV("INT pin defined but could not be initialized", dev);
+            return -MCP23X17_ERROR_INT_PIN;
+        }
+
+        /* create the interrupt handler thread */
+        if (_irq_thread_pid == KERNEL_PID_UNDEF) {
+            DEBUG("[%s] create irq_handler thread\n", __func__);
+            _irq_thread_pid = thread_create(_irq_thread_stack,
+                                            sizeof(_irq_thread_stack),
+                                            MCP23X17_IRQ_THREAD_PRIO,
+                                            THREAD_CREATE_STACKTEST,
+                                            _irq_thread, NULL, "irq_handler");
+            assert(_irq_thread_pid != KERNEL_PID_UNDEF);
+        }
+
+        /* initialize the IRQ event object used for delaying interrupts */
+        dev->irq_event.event.handler = _irq_handler;
+        dev->irq_event.dev = dev;
+
+        for (unsigned i = 0; i < MCP23X17_GPIO_PIN_NUM; i++) {
+            dev->isr[i].cb = NULL;
+            dev->isr[i].arg = NULL;
+        }
+
+
+        /* We use push/pull outputs and active low interrupt signals. */
+        iocon &= ~MCP23X17_IOCON_ODR;
+        iocon &= ~MCP23X17_IOCON_INTPOL;
+
+        /*
+         * Since we use only one pin for INTA and INTB signal, we have to use
+         * the MIRROR mode, i.e., an interrupt on either port will cause both
+         * interrupt pins to activate.
+         */
+        iocon |= MCP23X17_IOCON_MIRROR;
+
+        /*
+         * Reset all interrupt-on-change control bits to 0, that is, pin values
+         * are compared against the previous value for interrupt-on-change.
+         * Disable all interrupts. Since this corresponds to the power on reset
+         * configuration written before, we have not execute it here.
+         */
+        #if 0
+        uint8_t zero = 0;
+        if ((res = _write(dev, MCP23X17_REG_INTCONA, &zero, 1) ||
+            (res = _write(dev, MCP23X17_REG_INTCONB, &zero, 1) ||
+            (res = _write(dev, MCP23X17_REG_GPINTENA, &zero, 1) ||
+            (res = _write(dev, MCP23X17_REG_GPINTENB, &zero, 1)) {
+
+            DEBUG_DEV("error writing INTCON and GPINTEN registers", dev);
+            return res;
+        }
+        #endif
+    }
+#endif /* MODULE_PERIPH_GPIO_IRQ */
+
+    /* sequential operation and slew rate control for SDA used */
+    iocon &= ~MCP23X17_IOCON_SEQOP;
+    iocon &= ~MCP23X17_IOCON_DISSLW;
+
+    /* write back configuration registers */
+    if ((res = _write(dev, MCP23X17_REG_IOCONA, &iocon, 1)) ||
+        (res = _write(dev, MCP23X17_REG_IOCONB, &iocon, 1))) {
+        DEBUG_DEV("error writing IOCON registers", dev);
+        return res;
+    }
+
+    return MCP23X17_OK;
+}
+
+int mcp23x17_gpio_init(mcp23x17_t *dev, gpio_t pin, gpio_mode_t mode)
```

Comment:
I'm bit unsure about using `gpio_t` in here as some CPUs could define it as an structure, I think it may be more suitable to add a proper `mc23x17_gpio_t` type, same could apply for mode as CPUs could define their own and probably the IO expander has it's modes defined independently of the host.

Reply:
`gpio_t` and `gpio_mode_t` are used here intentionally. As a result of the review of the first driver provided in PR #7652, the goal of this PR was to implement an API that is compatible with the GPIO API in order to integrate the GPIO expanders transparently.

At the moment, gpio_t must be a scalar for all GPIO implementations because there are hundreds of comparisons in RIOT that use the == operator.

In PR #12877 and later in PR #14602 I tried to introduce a GPIO API using a structured `gpio_t` type more than a year ago, but I didn't succeed until now. So there is no reason not to use `gpio_t` at the moment. Once, the structured GPIO is introduced, the driver could easily be changed, as the [example of the low-level GPIO API implementation of an expander driver](https://github.com/RIOT-OS/RIOT/pull/14602/files#diff-f5b86696beed1342e9731fb44e4aa41f0c744de7d25c9ee5933731675734fe0c) in PR #14602 shows.